### PR TITLE
frontend tensorflow.nn convolution

### PR DESCRIPTION
**Pull request by hamehrabi**

### CODE REVIEW

**Comment by hamehrabi** [comment]
Hi!

Just for clarification of my codes:
There were two ways to implement this code. According to the TensorFlow document for 'tf.nn.convolution': (https://www.tensorflow.org/api_docs/python/tf/nn/convolution), this function computes sums of N-D convolutions-that is- it has similar functionality to Con1d, Conv2d, and Conv3d, based on its input dimensions. 
For example, if our input has a 4 dimension [batch, heights, widths, channels] this function will act like Conv2d.
As mentioned before there are two ways to implement this function:

**The First way (which is used in this pull request): Control the function based on data_format argument:**
In this method, since each of (conv1d, 2d, 3d) has its own specific 'data_format' string i.e:
            "conv1d: data_format ="NWC" (default) or "NCW"
            "conv2d= data_format = "NHWC" (default)  or "NCHW"
            "conv3d=  data_format = "NDHWC" (default)  or "NCDHW"

we can determine which ivy API function should be called. (For example if our data_format was==NWC, then our frontend function call "conv1d" from the ivy functional API.). This method assumed that data_format cannot be "None" and it should be determined by the user so that the function can call the relevant ivy functional API function.

Consequently, in our 'test_nn' section when we want to create a test function for convolution, based on an ivy API function which was called by 'data_format' arguments, we should choose the related test functions among our choices: 
i.e, "test_tensorflow_conv1d", "test_tensorflow_conv2d", "test_tensorflow_conv3d"

**The second way (TensorFlow used this method) is: Define the "convolution_internal" function and control the input dimensions based on this function.**
Link: (https://github.com/tensorflow/tensorflow/blob/v2.10.0/tensorflow/python/ops/nn_ops.py#L1171-L1188).
As far as I know, "convolution_internal" has been not implemented in ivy functional API and ivy TensorFlow backend. Therefore, it is recommended to implement this function first and then implement convolution based on this function.
Since this function has not existed in ivy API and backend, I decided to follow the first method.
(In this method, 'data_format' can be equal to "None" and the function automatically determine the type of function based on the input dimension.)

Please let me know, which method you prefer. 

Many thanks,
Hamed

**Comment by hirwa-nshuti** [comment]
HI @hamehrabi 
Thank you for your contribution😊

The first thing you have to do is commenting the issue this PR addresses by using close #issue_number.
About the implementation of a general convolution function on Ivy API, we currently have [conv_general_dilated](https://lets-unify.ai/ivy/functional/ivy/layers/conv_general_dilated.html) which can handle the convolution function to call depending on input dimensions.

Let me know if you have any questions.
Thanks🙂

**Comment by hamehrabi** [comment]
close #6736

**Comment by hamehrabi** [comment]
> HI @hamehrabi Thank you for your contribution😊
> 
> The first thing you have to do is commenting the issue this PR addresses by using close #issue_number. About the implementation of a general convolution function on Ivy API, we currently have [conv_general_dilated](https://lets-unify.ai/ivy/functional/ivy/layers/conv_general_dilated.html) which can handle the convolution function to call depending on input dimensions.
> 
> Let me know if you have any questions. Thanks🙂

Dear @hirwa-nshuti,

Thank you very much for your comment 😊👌. I modified my function and the test using the 'conv_general_dilated'. It seems to work, but I guess there is still a problem with 'conv_general_dilated'. 
In [convolution_internal](https://github.com/tensorflow/tensorflow/blob/v2.10.0/tensorflow/python/ops/nn_ops.py#L1171-L1188) [lines:1237-1252], there is an option to determine input rank even we do not use 'num_spatial_dims' -that is- the function automatically define the input rank, while I think in 'conv_general_dilated', dims is a mandatory argument and definitely should determine by the user so that function can determine the rank of input. 

In convolution_internal, the argument 'num_spatial_dims' can be None, and the function will find the rank, and based on the rank it will return conv1d, conv2d, and conv3d.
While, in conv_general_dilated, if the argument 'dims =None', the function assumes that dims=2 default and will not return any of the conv1d or conv3d functions. (please see ivy.functional.backends.tensorflow.layers--> conv_general_dilated)

I hope my explanation is clear.

I would be happy to know your opinion in this regard.

Many thanks,
Hamed

**Comment by hirwa-nshuti** [comment]
Hey @hamehrabi 

Yea your concerns make sense but I think this is a small use case that can be handled by a small helper block of code for the case where all other behavior is well fitting the implementation. And as in [superset behavior ](https://lets-unify.ai/ivy/deep_dive/superset_behaviour.html), some of the features are sacrificed during the implementation this might be the case.

I hope it is easy to implement this code block where the `num_spatial_dims` is set to `None`.

Best🙂

**Comment by hamehrabi** [comment]
> er block of code for the case where all other behavior is well fitting the implemen

Hey @hirwa-nshuti,

Thanks for your answer. Yea, you're totally right👌. It was my first try. May I know if my PR is now ready to merge or not? 

Best😊

File: ivy_tests/test_ivy/test_frontends/test_tensorflow/test_nn.py [hunk 1]
```diff
@@ -557,6 +557,36 @@ def test_tensorflow_conv3d_transpose(
         dilations=1 if not ivy.gpu_is_available() else dilation,
     )
 
+    
+@handle_cmd_line_args
+@given(
+    x_f_d_df=_x_and_filters(
+        dtypes=helpers.get_dtypes("float", full=False),
+        data_format=st.sampled_from(["NWC","NCW","NHWC","NCHW","NDHWC","NCDHW"]),
+        padding=st.sampled_from(["VALID", "SAME"]),
```
Comment (hirwa-nshuti) [inline ivy_tests/test_ivy/test_frontends/test_tensorflow/test_nn.py:566]: Here you will need to adjust the behavior of the `_x_and_filters` helper function to make sure that the data format selected matches the type of array being generated

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
The tests are failing now due to that there is no way to handle the type of generated arrays whether `1D` `2D` or `3D`
Some changes need to be made
Thanks🙂

**Comment by hamehrabi** [comment]
Dear @hirwa-nshuti,

After changing my code, it now passed all tests without any errors on my system. Could you please tell me if there are any errors to fix so I can merge the code?😊🙏

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Actually, the issue is still the arrays being generated you will need to modify the helper function to generate arrays suitable for the `convolution function` because the tests are failing on CI and even locally the tests will fail if you increase the number of testing examples. Have a look at failing tests from the checks tab to have a clear understanding of what's going wrong.
Best🙂

**Comment by hamehrabi** [comment]
Dear @hirwa-nshuti,

I hope you're doing well,

I have changed my codes multiple times, but the errors have remained. I just wondering what factors might be causing these errors.

**Problem:**
In this issue, we aim to join the "tf.nn.convolution" frontend function to the "conv_general_dilated()" backends' function by using the mediating role of the "ivy functional API function = conv_general_dilated()".

In our functional API and the backends, the function "conv_general_dilated()" accepts "dims" and "data_format" as the input arguments, and in fact, based on the "dims" they will determine the types of the output arrays. While in our frontend function **'tf.nn.convolution** only "data_format" can be accepted by the function. 

Also, "data_format" in the TensorFlow frontend accepts different inputs than ivy API and all backends. In other words:

1. [**Frontend function:**](https://www.tensorflow.org/api_docs/python/tf/nn/convolution) -->`tf.nn.convolution(data_format=['NWC', 'NCW', 'NHWC', 'NCHW', 'NDHWC', 'NCDHW'])`
2. [**ivy functional API & Backends functions:**](https://github.com/hamehrabi/ivy/blob/master/ivy/functional/ivy/layers.py#L1288-L1342) -->`ivy.conv_general_dilated(data_format=['channel_last', 'channel_first'])`

3. [**Frontend function:**](https://github.com/tensorflow/tensorflow/blob/v2.10.0/tensorflow/python/ops/nn_ops.py#L1171-L1188) -->`tf.nn.convolution(dims= No exist)`
4. [**ivy functional API & Backends functions:**](https://github.com/hamehrabi/ivy/blob/master/ivy/functional/ivy/layers.py#L1288-L1342) -->`ivy.conv_general_dilated(dims: int = 2)`

**Proposed Solution:** 
Since in the arguments 'data_format':
```
data_format('channel_last') = ['NWC', 'NHWC', 'NDHWC'] 
data_format('channel_first')  = ['NCW', 'NCHW', 'NCDHW']
```
**as well as based on the [TensorFlow documents](https://www.tensorflow.org/api_docs/python/tf/nn/convolution):** 
```
if data_format(['NWC', 'NCW']):
    dims = 1
elif data_format(['NHWC', 'NCHW']):
    dims = 2
elif data_format(['NDHWC', 'NCDHW']):
    dims = 3
```

I created two helper functions: 
The first one is:  ["_get_dims_channel()"](https://github.com/hamehrabi/ivy/blob/master/ivy_tests/test_ivy/test_frontends/test_tensorflow/test_nn.py#L217-L241)
This function has been created in order to convert the input values of "data_format=['NWC', 'NCW', 'NHWC', 'NCHW', 'NDHWC', 'NCDHW'])" to the output "dims" and "data_format", which will use as the inputs in ivy API and backends as well.
The second function is: ["_x_and_filters_convolution_channel_last()"](https://github.com/hamehrabi/ivy/blob/master/ivy_tests/test_ivy/test_frontends/test_tensorflow/test_nn.py#L244-L462)
This helper function will create the test samples and return those samples +(dims + data_format) which will pass through ivy API and the backends.

At the moment, I evaluated the functions in backends (tf, torch, NumPy, and Jax), and ivy API, but I am not sure what reasons caused these errors in our test results. I would be grateful if you look through my codes and let me know if you find any things can help to solve the problems. 

I am not sure but I think "_get_dims_channel()" should use in all the backends or ivy API as well. What do think about possible solutions? 😊🙄🛠

**Comment by hamehrabi** [comment]
reopen #6754

**Comment by hamehrabi** [comment]
Dear @hirwa-nshuti,

I hope you're doing well 😊
How can I re-open this issue? I wanted to make a pull request, but there was a problem with my fork, and I had to discard all the commits, today, I wanted to make a pull request again, but I couldn't re-open it. The icon has been disabled for me.

I also found some test functions in the repo that have problems passing the tests; how can I report them?

Best

**Comment by hirwa-nshuti** [comment]
Hi @hamehrabi 

To reopen this PR you need to revert the last changes you made and push back to the last version.
I'll also review it once you do it. Sorry for the delays I was not available during the weekends🙂

**Comment by hamehrabi** [comment]
> Hi @hamehrabi
> 
> To reopen this PR you need to revert the last changes you made and push back to the last version. I'll also review it once you do it. Sorry for the delays I was not available during the weekends🙂

Thank you for your reply. Sorry, if I bother you on the weekend. Unfortunately, I deleted that fork, and I do not have access to revert it, but in my new fork, I have done what I did before. So, do you think I have to create a new PR in this regard?

**Comment by hirwa-nshuti** [comment]
There is no problem in asking for reviews on Weekends I normally respond but for this one, I wasn't available at all.
Anyway go ahead and open a new PR🙂

**Pull request closed without merging**